planner_cspace/src/planner_3d.cpp

```diff
@@ -2051,6 +2067,174 @@ class Planner3dNode
+        // Update without region.
+        // Distance map will expand distance map using edges_buf if needed.
+        cost_estim_cache_static_.update(
+            s, metric2Grid(goal_original_.pose),
+            DistanceMap::Rect(Astar::Vec(1, 1, 0), Astar::Vec(0, 0, 0)));
+        const auto tnow = boost::chrono::high_resolution_clock::now();
+        const float dur = boost::chrono::duration<float>(tnow - ts).count();
+        ROS_DEBUG("Cost estimation cache for static map updated (%0.4f sec.)", dur);
+        metrics_.data.push_back(neonavigation_metrics_msgs::metric(
+            "distance_map_static_update_dur", dur, "second"));
+        metrics_.data.push_back(neonavigation_metrics_msgs::metric(
+            "distance_map_static_init_dur", 0.0, "second"));
+      }
+
+      // Construct arraivability map
+      const int local_width = esc_range_ * 2 + 1;
+      const Astar::Vec local_origin = s - Astar::Vec(esc_range_, esc_range_, 0);
+      const Astar::Vec local_size(local_width, local_width, 1);
+      const Astar::Vec local_center(esc_range_, esc_range_, 0);
+      const DistanceMap::Params dmp =
+          {
+              .euclid_cost = ec_,
+              .range = 0,
+              .local_range = 0,
+              .longcut_range = esc_range_,
+              .size = local_size,
+              .resolution = map_info_.linear_resolution,
+          };
+
+      cm_local_esc_.reset(local_size);
+      arrivable_map_.setParams(cc_, num_cost_estim_task_);
+      arrivable_map_.init(model_, dmp);
+      cm_local_esc_.copy_partially(
+          Astar::Vec(0, 0, 0), cm_rough_, local_origin, local_origin + local_size);
+      arrivable_map_.create(local_center, local_center);
+
+      // Find temporary goal
+      float cost_min = std::numeric_limits<float>::max();
+      Astar::Vec te_out;
+      const int esc_range_sq = esc_range_ * esc_range_;
+      const int esc_range_min_sq = esc_range_min_ * esc_range_min_;
+      for (Astar::Vec d(0, -esc_range_, 0); d[1] <= esc_range_; d[1]++)
+      {
+        for (d[0] = -esc_range_; d[0] <= esc_range_; d[0]++)
+        {
+          const int sqlen = d.sqlen();
+          if ((d[0] == 0 && d[1] == 0) || sqlen > esc_range_sq || sqlen < esc_range_min_sq)
+          {
+            continue;
+          }
+
+          Astar::Vec te(start_grid[0] + d[0], start_grid[1] + d[1], 0);
+          if ((unsigned int)te[0] >= (unsigned int)map_info_.width ||
+              (unsigned int)te[1] >= (unsigned int)map_info_.height)
+          {
+            continue;
+          }
+          if (!cm_rough_.validate(te, range_) || cm_rough_[te] >= 50)
+          {
+            continue;
+          }
+          const auto cost = cost_estim_cache_static_[te];
+          if (cost >= cost_min)
+          {
+            continue;
+          }
+
+          // Check arrivability
+          if (arrivable_map_[te - local_origin] == std::numeric_limits<float>::max())
+          {
+            continue;
+          }
+
```

Comment:
Need to go back to the normal goal if the original goal is one of the candidate of temporary escape